src/util/futurepool.rs

```diff
@@ -0,0 +1,259 @@
+/// A delegator to wrap `Context` to provide `on_tick` feature.
+struct ContextDelegator<T: Context> {
+    tick_interval: Duration,
+    /// TODO: Can be replace by `UnsafeCell` to eliminate the runtime cost of borrow checking.
+    inner: RefCell<T>,
+    last_tick: Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextDelegator<T> {
+    fn new(context: T, tick_interval: Duration) -> ContextDelegator<T> {
+        ContextDelegator {
+            tick_interval,
+            inner: RefCell::new(context),
+            last_tick: Cell::new(None),
+        }
+    }
+
+    fn get_context(&self) -> RefMut<T> {
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.get_context().on_tick();
+    }
+}
+
+/// Each `ContextDelegator` instance is invoked individually for each thread
+/// so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextDelegator<T> {}
+
+#[derive(Debug)]
+pub struct ContextDelegators<T: Context> {
+    delegators: Arc<HashMap<thread::ThreadId, ContextDelegator<T>>>,
+}
+
+impl<T: Context> Clone for ContextDelegators<T> {
+    fn clone(&self) -> Self {
+        ContextDelegators::<T> {
+            delegators: Arc::clone(&self.delegators),
+        }
+    }
+}
+
+impl<T: Context> ContextDelegators<T> {
+    fn new(delegators: HashMap<thread::ThreadId, ContextDelegator<T>>) -> Self {
+        ContextDelegators::<T> {
+            delegators: Arc::new(delegators),
+        }
+    }
+
+    fn get_current_thread_delegator(&self) -> &ContextDelegator<T> {
+        let thread_id = thread::current().id();
+        if let Some(delegator) = self.delegators.get(&thread_id) {
+            delegator
+        } else {
+            unreachable!();
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A future thread pool that supports `on_tick` for each thread.
+pub struct FuturePool<T: Context> {
+    pool: cpupool::CpuPool,
+    context_delegators: ContextDelegators<T>,
+}
+
+impl<T: Context> Clone for FuturePool<T> {
+    fn clone(&self) -> FuturePool<T> {
+        FuturePool::<T> {
+            pool: self.pool.clone(),
+            context_delegators: self.context_delegators.clone(),
+        }
+    }
+}
+
+impl<T: Context> util::AssertSend for FuturePool<T> {}
+impl<T: Context> util::AssertSync for FuturePool<T> {}
+
+impl<T: Context> FuturePool<T> {
+    pub fn new<F>(
+        pool_size: usize,
+        stack_size: usize,
+        name_prefix: &str,
+        tick_interval: Duration,
+        context_factory: F,
+    ) -> FuturePool<T>
+    where
+        F: Send + 'static + Fn(thread::ThreadId) -> T,
```

Comment:
Why we need `ThreadId` as argument? Any example?

Reply:
Might be useful so that it is provided anyway. For example, we might want to store it in the context to help distinguish which thread this context belongs to to help debug.

Reply:
Can it be obtained by simply `thread::current().id()`?

Reply:
@BusyJay Nice!